File: backend/app/models/menu_config.py
"""
選單設定 Models
- MenuConfig        : 每個 menu key 的自訂 label 與排序
- MenuConfigHistory : 每次 Save 的快照，最多保留 5 筆
"""
from sqlalchemy import String, Integer, Boolean, DateTime, Text
from sqlalchemy.orm import Mapped, mapped_column
from datetime import datetime
import uuid
from app.core.database import Base
from app.core.time import twnow


def _now():
    return twnow()


class MenuConfig(Base):
    __tablename__ = "menu_configs"

    # menu key：一級群組用群組 key（如 'hotel'），二級頁面用路由（如 '/hotel/periodic-maintenance'）
    menu_key: Mapped[str] = mapped_column(String(120), primary_key=True)
    # 父層 key；一級群組為 None
    parent_key: Mapped[str | None] = mapped_column(String(120), nullable=True, default=None)
    # 自訂顯示名稱；None 表示使用 navLabels 預設值
    custom_label: Mapped[str | None] = mapped_column(String(100), nullable=True, default=None)
    # 排序順序（同層內由小到大）
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    # 保留欄位：未來可做隱藏功能
    is_visible: Mapped[bool] = mapped_column(Boolean, default=True)

    updated_at: Mapped[datetime] = mapped_column(DateTime, default=_now, onupdate=_now)
    updated_by: Mapped[str | None] = mapped_column(String(100), nullable=True, default=None)


class MenuConfigHistory(Base):
    __tablename__ = "menu_config_history"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    changed_at: Mapped[datetime] = mapped_column(DateTime, default=_now)
    changed_by: Mapped[str] = mapped_column(String(100), nullable=False)
    # 本次變更摘要（JSON 字串）：記錄哪些 key 的 label / order 發生了變化
    diff_json: Mapped[str] = mapped_column(Text, nullable=False)
    # 儲存當下全量快照（JSON 字串），可供還原參考
    snapshot_json: Mapped[str] = mapped_column(Text, nullable=False)

File: frontend/src/api/menuConfig.ts
/**
 * 選單設定 API 封裝
 * 對應後端 /settings/menu-config
 */
import apiClient from '@/api/client'

export interface MenuConfigItem {
  menu_key: string
  parent_key: string | null
  custom_label: string | null
  sort_order: number
  is_visible: boolean
  // 權限控制：null = 公開顯示；有值 = 需具備對應 permission_key
  permission_key: string | null
}

export interface MenuConfigHistoryItem {
  id: string
  changed_at: string
  changed_by: string
  diff_json: string     // JSON string：[{key, label?, order?}, ...]
  snapshot_json: string // JSON string：MenuConfigItem[]
}

/** 取得目前全部設定 */
export async function fetchMenuConfig(): Promise<MenuConfigItem[]> {
  const res = await apiClient.get<MenuConfigItem[]>('/settings/menu-config')
  return res.data
}

/** 批次儲存設定（覆寫全部） */
export async function saveMenuConfig(items: MenuConfigItem[]): Promise<MenuConfigItem[]> {
  const res = await apiClient.put<MenuConfigItem[]>('/settings/menu-config', { items })
  return res.data
}

/** 取得最近 5 筆變更記錄 */
export async function fetchMenuConfigHistory(): Promise<MenuConfigHistoryItem[]> {
  const res = await apiClient.get<MenuConfigHistoryItem[]>('/settings/menu-config/history')
  return res.data
}
